File: aisd/src/main/java/org/example/structure/graph/Graph.java
package org.example.structure.graph;

import java.util.*;

// Граф, алгоритм обхода в ширину, алгоритм обхода в длину
public class Graph {
    private static class Vertex {
        private int id;
        private List<Edge> edges = new ArrayList<>();

        public Vertex(int id) {
            this.id = id;
        }

        public int getId() {
            return id;
        }

        public List<Edge> getEdges() {
            return edges;
        }

        public void addEdge(Edge edge) {
            edges.add(edge);
        }

        @Override
        public String toString() {
            return "Element " + id;
        }
    }

    private static class Edge {
        private Vertex vertex;

        public Edge(Vertex vertex) {
            this.vertex = vertex;
        }

        public Vertex getVertex() {
            return vertex;
        }
    }

    // List to store vertices
    private List<Vertex> vertices = new ArrayList<>();

    // Method to add a new vertex
    public void addVertex() {
        vertices.add(new Vertex(vertices.size()));
    }

    // Method to get a vertex by id
    public Vertex getVertex(int id) {
        for (Vertex vertex : vertices) {
            if (vertex.getId() == id) {
                return vertex;
            }
        }
        return null;
    }

    // Method to add an edge between two vertices
    public void addEdge(int id1, int id2) {
        Vertex vertex1 = getVertex(id1);
        Vertex vertex2 = getVertex(id2);
        if (vertex1 != null && vertex2 != null) {
            vertex1.addEdge(new Edge(vertex2));
            vertex2.addEdge(new Edge(vertex1));
        }
    }

    // Method to perform Breadth-First Search starting from a given vertex id
    public void bfs(int id) {
        int[] array = new int[vertices.size()];
        for (int i = 0; i < array.length; i++) {
            array[i] = -1;
        }
        Queue<Vertex> queue = new LinkedList<>();

        queue.add(getVertex(id));

        int iteration = 1;
        while (!queue.isEmpty()) {
            Vertex vertex = queue.poll();

            System.out.println("Iteration " + iteration + ": Algorithm entered vertex " + vertex.getId());
            array[vertex.getId()] = iteration;

            List<Edge> edges = vertex.getEdges();
            for (Edge edge : edges) {
                if (array[edge.getVertex().getId()] == -1 && !queue.contains(getVertex(edge.getVertex().getId()))) {
                    System.out.println("Algorithm added vertex " + edge.getVertex().getId() + " to the queue");
                    queue.add(edge.getVertex());
                }
            }
            iteration++;
        }

        System.out.println("Algorithm traversed all vertices");
    }

    // Method to perform Depth-First Search starting from a given vertex id
    public void dfs(int startId) {
        boolean[] visited = new boolean[vertices.size()];
        var vertex = getVertex(startId);
        Stack<Vertex> stack = new Stack<>();
        stack.push(vertex);

        while (!stack.isEmpty()) {
            Vertex currentVertex = stack.pop();

            if (!visited[currentVertex.getId()]) {
                visited[currentVertex.getId()] = true;
                System.out.println("Visited vertex: " + currentVertex.getId());

                // Process adjacent vertices
                for (Edge edge : currentVertex.getEdges()) {
                    Vertex adjacentVertex = edge.getVertex();
                    if (!visited[adjacentVertex.getId()]) {
                        stack.push(adjacentVertex);
                    }
                }
            }
        }
    }

    // Method to generate and print adjacency list
    public void printAdjacencyList() {
        for (Vertex vertex : vertices) {
            System.out.print("Vertex " + vertex.getId() + " -> ");
            for (Edge edge : vertex.getEdges()) {
                System.out.print(edge.getVertex().getId() + " ");
            }
            System.out.println();
        }
    }

    // Method to generate and print adjacency matrix
    public void printAdjacencyMatrix() {
        int[][] adjacencyMatrix = new int[vertices.size()][vertices.size()];
        for (Vertex vertex : vertices) {
            for (Edge edge : vertex.getEdges()) {
                adjacencyMatrix[vertex.getId()][edge.getVertex().getId()] = 1;
            }
        }

        for (int i = 0; i < vertices.size(); i++) {
            for (int j = 0; j < vertices.size(); j++) {
                System.out.print(adjacencyMatrix[i][j] + " ");
            }
            System.out.println();
        }
    }

    // Method to generate and print incidence matrix
    public void printIncidenceMatrix() {
        int[][] incidenceMatrix = new int[vertices.size()][vertices.size()];
        for (Vertex vertex : vertices) {
            for (Edge edge : vertex.getEdges()) {
                incidenceMatrix[vertex.getId()][edge.getVertex().getId()] = 1;
                incidenceMatrix[edge.getVertex().getId()][vertex.getId()] = 1;
            }
        }

        for (int i = 0; i < vertices.size(); i++) {
            for (int j = 0; j < vertices.size(); j++) {
                System.out.print(incidenceMatrix[i][j] + " ");
            }
            System.out.println();
        }
    }
}

File: misc/ASD/ASD/Graph.cs
namespace ASD;

public class Graph
{
    public List<Vertex> Vertex { get; set; } = new List<Vertex>();

    public void AddVertex()
    {
        Vertex.Add(new Vertex { Id = Vertex.Count });
    }

    public Vertex? GetVertex(int id)
    {
        return Vertex.FirstOrDefault(x => x.Id == id);
    }

    public void AddEdge(int id1, int id2)
    {
        var vertex1 = GetVertex(id1);
        var vertex2 = GetVertex(id2);
        if (vertex1 != null && vertex2 != null)
        {
            vertex1.AddEdge(new Edge { Vertex = vertex2 });
            vertex2.AddEdge(new Edge { Vertex = vertex1 });
        }
    }

    public void BFS(int id)
    {
        var array = new int[Vertex.Count];
        for (int i = 0; i < array.Length; i++)
        {
            array[i] = -1;
        }
        var queue = new Queue<Vertex>();

        queue.Enqueue(Vertex.First(x => x.Id == id));

        int iteration = 1;
        while (queue.Count != 0)
        {
            var vertex = queue.Dequeue();

            Console.WriteLine($"Итерация {iteration}: Алгоритм зашел в {vertex.Id} вершину");
            array[vertex.Id] = iteration;

            var edges = vertex.Edges;
            foreach (var edge in edges)
            {
                if (array[edge.Vertex.Id] == -1 && !queue.Contains(GetVertex(edge.Vertex.Id)))
                {
                    Console.WriteLine($"Алгоритм добавил в очередь {edge.Vertex.Id} вершину");
                    queue.Enqueue(edge.Vertex);
                }
            }
            iteration++;
        }

        Console.WriteLine($"Алгоритм прошел все вершины");
    }
}

public class Vertex
{
    public int Id { get; set; }

    public List<Edge> Edges { get; set; } = new List<Edge>();

    public void AddEdge(Edge edge)
    {
        Edges.Add(edge);
    }

    public override string ToString()
    {
        return $"Элемент {Id}";
    }
}

public class Edge
{
    public Vertex Vertex { get; set; }

}